src/Mutation/PrioritizedVisitorsNodeTraverser.php

```diff
@@ -0,0 +1,123 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Mutation;
+
+use DomainException;
+use function get_class;
+use InvalidArgumentException;
+use function krsort;
+use PhpParser\Node;
+use PhpParser\NodeTraverserInterface;
+use PhpParser\NodeVisitor;
+use function Safe\sprintf;
+use const SORT_NUMERIC;
+
+/**
+ * @internal
+ * @final
+ */
+class PrioritizedVisitorsNodeTraverser implements NodeTraverserInterface
+{
+    /**
+     * @var array<int,NodeVisitor>
+     */
+    private $visitors = [];
+
+    private $traverser;
+
+    public function __construct(NodeTraverserInterface $decoratedTraverser)
+    {
+        $this->traverser = $decoratedTraverser;
+    }
+
+    public function addPrioritizedVisitor(NodeVisitor $visitor, int $priority): void
+    {
+        if (array_key_exists($priority, $this->visitors)) {
+            throw new InvalidArgumentException(sprintf(
+                'The priority "%d" is already used for the visitor "%s". Please use a different one',
+                $priority,
+                get_class($this->visitors[$priority])
+            ));
+        }
+
+        $this->visitors[$priority] = $visitor;
+    }
+
+    public function addVisitor(NodeVisitor $visitor): void
+    {
+        throw new DomainException('Add a non-prioritized visitor is not supported.');
+    }
+
+    public function removeVisitor(NodeVisitor $visitor): void
+    {
+        throw new DomainException('Removing a visitor is not supported.');
+    }
+
+    /**
+     * @return array<int,NodeVisitor>
+     */
+    public function getVisitors(): array
+    {
+        // This could on theory be optimized by doing it only once. This would however require to
+        // keep track of whether or not visitors have been already sorted and reset that state
+        // when adding a visitor.
+        // Since this method is mostly used for test-purposes to inspect the configured visitors,
+        // it is most definitely not worth to add more state for this nano-optimization.
+        krsort($this->visitors, SORT_NUMERIC);
+
+        return $this->visitors;
+    }
+
+    /**
+     * @return Node[]
+     */
+    public function traverse(array $nodes): array
+    {
+        // This could on theory be optimized by doing it only once. This would however require to
+        // keep track of whether or not visitors have been already sorted and reset that state
+        // when adding a visitor.
+        // Since in practice this method is used only once this would be within the real of
+        // nano-optimization hence not worth pursuing since would introduce more state and make
+        // the tests of this class unnecessarily more complex.
+        krsort($this->visitors, SORT_NUMERIC);
```

Comment:
if this is really a "nano-optimization" as you say, would it make sense to move it back to `addPrioritizedVisitor()` and don't have this sorting in 2 places with these long comments? :)

Personally, I think we are ok to call `krsort` 5-10 times while mutation testing can take hours

Reply:
I guess yeah